RESTEasy tries to instantiate abstract resource base class

**Describe the bug**
If your REST resource extends an abstract base class that contains all the typical annotations like `@Path`, `@GET` etc. then you'll see this failure when invoking the resource:
```
2020-10-25 23:39:01,178 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /hello failed, error id: fb45d7cc-5fa1-4aa5-ae5c-fa458d3b702d-1: org.jboss.resteasy.spi.UnhandledException: java.lang.RuntimeException: java.lang.InstantiationException
        at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:131)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:37)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:94)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: java.lang.RuntimeException: java.lang.InstantiationException
        at io.quarkus.arc.runtime.BeanContainerImpl$DefaultInstanceFactory.create(BeanContainerImpl.java:75)
        at io.quarkus.resteasy.common.runtime.QuarkusConstructorInjector.construct(QuarkusConstructorInjector.java:54)
        at org.jboss.resteasy.plugins.server.resourcefactory.POJOResourceFactory.createResource(POJOResourceFactory.java:71)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:386)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        ... 17 more
Caused by: java.lang.InstantiationException
        at java.base/jdk.internal.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at io.quarkus.arc.runtime.BeanContainerImpl$DefaultInstanceFactory.create(BeanContainerImpl.java:67)
        ... 22 more
```
Such abstract base classes can be the result of https://github.com/OpenAPITools/openapi-generator

**Expected behavior**
Resource classes can use abstract base classes (that contain JAX-RS annotations) without problems.

**Actual behavior**
Invocation of such a resource fails.

**To Reproduce**

Steps to reproduce the behavior:
1. `git clone https://github.com/famod/quarkus-quickstarts`
2. `git switch rest-abstract`
3. `cd getting-started`
4. `mvn clean verify` (will fail with the mentioned error)
5. `mvn clean package -DskipTests`
6. `java -jar target/getting-started-1.0-SNAPSHOT-runner.jar`
7. `curl localhost:8080/hello` (will fail with the mentioned error)

**Configuration**
not relevant

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `MINGW64_NT-10.0-18363 XXX 3.0.7-338.x86_64 2019-11-21 23:07 UTC x86_64 Msys`
 - Output of `java -version`: `OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)`
 - GraalVM version (if different from Java): n/a
 - Quarkus version or git rev: 1.9.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Apache Maven 3.6.3

**Additional context**
Interfaces do work but annotation discovery has some blind spots in that case:
- https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/jwt.20role.20handling
- #12736

--- a/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
+++ b/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
@@ -244,6 +244,7 @@ public class ResteasyServerCommonProcessor {
 
         Map<DotName, ClassInfo> scannedResources = new HashMap<>();
         Set<DotName> pathInterfaces = new HashSet<>();
+        Set<DotName> pathAbstract = new HashSet<>();
         Map<DotName, ClassInfo> withoutDefaultCtor = new HashMap<>();
         for (AnnotationInstance annotation : allPaths) {
             if (annotation.target().kind() == AnnotationTarget.Kind.CLASS) {
@@ -252,7 +253,11 @@ public class ResteasyServerCommonProcessor {
                     if (!withoutDefaultCtor.containsKey(clazz.name())) {
                         String className = clazz.name().toString();
                         if (!additionalPaths.contains(annotation)) { // scanned resources only contains real JAX-RS resources
-                            scannedResources.putIfAbsent(clazz.name(), clazz);
+                            if (Modifier.isAbstract(clazz.flags())) {
+                                pathAbstract.add(clazz.name());
+                            } else {
+                                scannedResources.putIfAbsent(clazz.name(), clazz);
+                            }
                         }
                         reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, className));
 
@@ -273,6 +278,21 @@ public class ResteasyServerCommonProcessor {
                 String className = implementor.name().toString();
                 reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, className));
                 scannedResources.putIfAbsent(implementor.name(), implementor);
+
+                if (!implementor.hasNoArgsConstructor()) {
+                    withoutDefaultCtor.put(implementor.name(), implementor);
+                }
+            }
+        }
+        // look for all implementations of abstract classes annotated @Path
+        for (final DotName cls : pathAbstract) {
+            final Collection<ClassInfo> implementors = index.getAllKnownSubclasses(cls);
+            for (final ClassInfo implementor : implementors) {
+                String className = implementor.name().toString();
+                reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, className));
+                if (!Modifier.isAbstract(implementor.flags())) {
+                    scannedResources.putIfAbsent(implementor.name(), implementor);
+                }
 
                 if (!implementor.hasNoArgsConstructor()) {
                     withoutDefaultCtor.put(implementor.name(), implementor);

--- a/integration-tests/resteasy-jackson/src/main/java/io/quarkus/it/resteasy/jackson/AbstractResource.java
+++ b/integration-tests/resteasy-jackson/src/main/java/io/quarkus/it/resteasy/jackson/AbstractResource.java
@@ -1,0 +1,18 @@
+package io.quarkus.it.resteasy.jackson;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+@Path("/abstract")
+public abstract class AbstractResource {
+
+    @GET
+    @Path("inherited")
+    @Produces("text/plain")
+    public String inherited() {
+        return message();
+    }
+
+    public abstract String message();
+}

--- a/integration-tests/resteasy-jackson/src/main/java/io/quarkus/it/resteasy/jackson/ConcreteClass.java
+++ b/integration-tests/resteasy-jackson/src/main/java/io/quarkus/it/resteasy/jackson/ConcreteClass.java
@@ -1,0 +1,8 @@
+package io.quarkus.it.resteasy.jackson;
+
+public class ConcreteClass extends AbstractResource {
+    @Override
+    public String message() {
+        return "concrete";
+    }
+}

--- a/integration-tests/resteasy-jackson/src/test/java/io/quarkus/it/resteasy/jackson/GreetingResourceTest.java
+++ b/integration-tests/resteasy-jackson/src/test/java/io/quarkus/it/resteasy/jackson/GreetingResourceTest.java
@@ -20,4 +20,14 @@ class GreetingResourceTest {
                 .body(containsString("[2019,1,1]"));
     }
 
+    @Test
+    public void testAbstract() {
+        given()
+                .when().get("/abstract/inherited")
+                .then()
+                .statusCode(200)
+                .body(containsString("concrete"));
+
+    }
+
 }

